In [ ]:
#hiddencell
from pbl_tools import *

# 스테이지 5

# 1. pandas를 이용해 csv 파일 읽어오기
[문제 1]  
Pandas 라이브러리(library)를 가져와보세요.   
그리고 train.csv, test.csv, sample_submission.csv를 각각     
raw_data_train, raw_data_test, raw_data_submission 변수로 할당하여 데이터를 읽어와 보세요.  

In [ ]:
import ___ as pd

raw_data_train = pd.___('train.csv')
raw_data_test = pd.___('test.csv')
raw_data_submission = pd.___('sample_submission.csv')

In [ ]:
#checkcode
ensure_vals(globals(), 'raw_data_train', 'raw_data_test', 'raw_data_submission')
@check_safety
def check(user_answer_train = raw_data_train, 
          user_answer_test = raw_data_test, 
          user_answer_submission = raw_data_submission
          ):
    
    c_point0 = hasattr(user_answer_train, 'tail')
    c_point1 = hasattr(user_answer_test, 'tail')
    c_point2 = hasattr(user_answer_submission, 'tail')

    if c_point0 and c_point1 and c_point2 :
        return True
    else:
        return False

check()

True

### Inst.  


### Hint.
`pd.read_csv('파일경로/파일명')` 를 활용해 보세요!

### Solution.
```python
import pandas as pd

raw_data_train = pd.read_csv('train.csv')  
raw_data_test = pd.read_csv('test.csv')  
raw_data_submission = pd.read_csv('sample_submission.csv') 
```

# 2. 원본 데이터 복사하여 train, test, submission 데이터셋 생성하기

[문제 2]   
`copy()` 메서드를 사용하여 raw_data_train, raw_data_test, raw_data_submission 원본 데이터를 복사해 보세요.

In [ ]:
train = raw_data_train.___()
test = raw_data_test.___()
submission = raw_data_submission.___()

In [ ]:
#checkcode
ensure_vals(globals(), 'train', 'test', 'submission')
@check_safety
def check(user_answer_train = raw_data_train, 
          user_answer_test = raw_data_test, 
          user_answer_submission = raw_data_submission
          ):
    
    c_point0 = hasattr(train, 'tail')
    c_point1 = hasattr(test, 'tail')
    c_point2 = hasattr(submission, 'tail')

    if c_point0 and c_point1 and c_point2 :
        return True
    else:
        return False

check()

### Inst.
해당 코드는 원본 데이터를 복사하여 새로운 데이터프레임(DataFrame)인 train, test, submission을 생성하는 코드입니다.

이 코드는 **원본 데이터를 보존하면서 독립적으로 작업** 하기 위해 사용됩니다.   
원본 데이터를 직접 수정하는 대신, 복사하여 새로운 데이터프레임을 생성하면 **원본 데이터의 무결성을 보존** 할 수 있습니다.   
이후의 데이터 전처리나 모델 학습 등의 작업에서는 이러한 새로운 데이터프레임을 사용할 수 있습니다.

`copy()` 메서드를 사용하여 원본 데이터를 복사하는 방법은 데이터프레임의 독립성을 유지하고, 원본 데이터와의 연관성을 끊을 수 있습니다.   
이를 통해 원본 데이터를 보존하면서 분석 및 처리 작업을 수행할 수 있습니다.


### Hint.
empty

### Solution.
```python
train = raw_data_train.copy()
test = raw_data_test.copy()
submission = raw_data_submission.copy()
```

# 3. train 데이터셋의 피처(feature)별 결측치 개수 확인하기 (1)

[문제 3]  
train 데이터셋의 피처(feature)별 결측치 개수의 합을 구해 보세요.  
아래 빈칸을 채워보세요.  

In [ ]:
train.___().sum()

title               0
distributor         0
genre               0
release_time        0
time                0
screening_rat       0
director            0
dir_prev_bfnum    330
dir_prev_num        0
num_staff           0
num_actor           0
box_off_num         0
dtype: int64

In [ ]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer=_):
    c_point0 = _[0] == 0
    if c_point0:
        return True
    else:
        return False
check()

True

### Inst.

### hint.
`isnull().sum()` 을 이용하면 train 데이터셋의 피처(feature)별 결측치의 수를 확인할 수 있습니다.

### solution.
```python
train.isnull().sum()
```

# 4. 결측치가 제거된 train_dropna 데이터셋 생성하기(dropna)

[문제 4]  
`dropna()` 함수를 이용하여 결측치가 포함된 행을 삭제해 보세요.


In [ ]:
train_dropna = train.___()
train_dropna

title   distributor genre release_time  time screening_rat  \
1              내부자들        (주)쇼박스   느와르   2015-11-19   130      청소년 관람불가   
2         은밀하게 위대하게        (주)쇼박스    액션   2013-06-05   123       15세 관람가   
3          나는 공무원이다        (주)NEW   코미디   2012-07-12   101        전체 관람가   
4              불량남녀  쇼박스(주)미디어플렉스   코미디   2010-11-04   108       15세 관람가   
5    강철대오 : 구국의 철가방      롯데엔터테인먼트   코미디   2012-10-25   113       15세 관람가   
..              ...           ...   ...          ...   ...           ...   
587         서유기 리턴즈     (주) 영화사조제    SF   2011-02-17    79        전체 관람가   
588             하울링     CJ 엔터테인먼트   느와르   2012-02-16   114       15세 관람가   
592           차이나타운      CGV아트하우스   느와르   2015-04-29   110      청소년 관람불가   
595              해무        (주)NEW   드라마   2014-08-13   111      청소년 관람불가   
596            파파로티        (주)쇼박스   드라마   2013-03-14   127       15세 관람가   

    director  dir_prev_bfnum  dir_prev_num  num_staff  num_actor  box_off_num  
1        우민호    1.161602e+06             2        387          3      7072501  
2        장철수    2.207752e+05             4        343          4      6959083  
3        구자홍    2.389400e+04             2         20          6       217866  
4        신근호    1.000000e+00             1        251          2       483387  
5        육상효    8.379690e+05             2        262          4       233211  
..       ...             ...           ...        ...        ...          ...  
587      신재호    6.760200e+04             1        220          4        12696  
588       유하    2.061297e+06             3        288          2      1612554  
592      한준희    5.974970e+05             3        420          3      1472147  
595      심성보    3.833000e+03             1        510          7      1475091  
596      윤종찬    4.960610e+05             1        286          6      1716438  

[270 rows x 12 columns]

In [ ]:
#checkcode
ensure_vals(globals(),'train','train_dropna')
@check_safety
def check(user_answer=_, user_train=train_dropna):

    c_point0 = (user_answer.shape == user_train.shape)

    if c_point0 :
        return True
    else:
        return False
check()

### Inst.
train 데이터프레임(DataFrame)에서 `dir_prev_bfnum` 피처(feature)에 약 330개의 결측치가 있는 것을 확인했습니다.  
결측치가 많은 열(column)을 그대로 두면 모델 학습에 방해가 될 수 있기 때문에,  
이러한 **결측치를 처리**해 주는 것이 중요합니다.  

`dropna()` 함수는 이러한 결측치 처리 과정 중 하나로, **결측치가 포함된 행을 삭제** 하는 방법입니다.  
해당 코드는 결측치가 포함된 train 데이터프레임에서 **결측치가 없는 행들만으로 이루어진 train_dropna** 데이터프레임을 반환합니다.  


<img width=585 src="https://drive.google.com/uc?id=1gGs3FVCc4iIxILeoZOFUseLmZ6ZcNgHE">


결측치가 포함된 행을 삭제하면 **데이터셋의 크기가 줄어들기** 때문에, 이에 따른 모델의 성능 변화도 고려해야 합니다.  
또한, 이 외에도 다양한 결측치 처리 방법들이 존재하므로, 상황에 맞는 적절한 결측치 처리 방법을 선택하는 것이 중요합니다.

### hint.
empty

### solution.
```python
train_dropna = train.dropna()
train_dropna
```

# 5. train_dropna 데이터셋의 피처(feature)별 결측치 개수 확인하기 (2)

[문제 5]  

결측치가 제대로 제거되었는지 확인해 봐야겠죠?  
피처별 결측치 개수의 합을 구해 보세요.


train_dropna 데이터셋의 피처(feature)별 결측치 개수의 합을 구해 보세요.  
아래 빈칸을 채워보세요.   

In [ ]:
train_dropna.___().___()

title             0
distributor       0
genre             0
release_time      0
time              0
screening_rat     0
director          0
dir_prev_bfnum    0
dir_prev_num      0
num_staff         0
num_actor         0
box_off_num       0
dtype: int64

In [ ]:
#checkcode
ensure_vals(globals(), 'train_dropna')
@check_safety
def check(check_train = train_dropna, user_answer = _):
    c_point0 = (user_answer == check_train.isnull().sum()).all()
    if c_point0:
        return True
    else:
        return False
check()

True

### Inst.


### Hint.
`isnull().sum()` 을 이용하면 train_dropna 데이터셋의 피처(feature)별 결측치의 수를 확인할 수 있습니다.

### Solution.
```python
train_dropna.isnull().sum()
```

# 6. 결측치가 제거된 train_dropna 데이터셋의 크기 확인하기

[문제 6]  
결측치들을 제거한 후의 행(row)과 열(column)의 크기를 확인해 봅시다.  
아래 빈칸을 채워주세요.

In [ ]:
train_dropna.___

(270, 12)

In [ ]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer=_,user_answer_dropna=train_dropna):
    c_point0 = user_answer == user_answer_dropna.shape
    if c_point0:
        return True
    else:
        return False
check()

True

### Inst.
`shape` 을 사용하여 데이터셋의 구조를 확인하고, 데이터셋이 어떤 형태로 구성되어 있는지 파악할 수 있습니다.

위에서 출력된 결과를 보니 모든 피처(feature)에 대해 결측치가 없는 것으로 나타났습니다.  
따라서 결측치 제거가 확실하게 이루어졌다는 것을 확인할 수 있습니다.

train 데이터프레임(DataFrame)은 (600, 12) 크기를 가졌지만,   
결측치가 존재했던 **330개의 행이 제거된 train_dropna 데이터프레임은 (270, 12)** 크기를 가지게 됩니다.


### Hint.
`shape` 함수를 사용해서 데이터프레임(dataframe)의 행(row)과 열(column)의 크기를 확인해 보세요.

### Solution.
```python
train_dropna.shape
```

# 7. 결측치를 0으로 채운 train_fill 데이터셋 생성하기

[문제 7]    
`fillna()` 메소드를 사용하여 pandas 데이터프레임(DataFrame)에서 결측치(missing value)를 0으로 채워보세요.

In [ ]:
train_fill = train.___(0)
train_fill

title   distributor genre release_time  time screening_rat director  \
0       개들의 전쟁      롯데엔터테인먼트    액션   2012-11-22    96      청소년 관람불가      조병옥   
1         내부자들        (주)쇼박스   느와르   2015-11-19   130      청소년 관람불가      우민호   
2    은밀하게 위대하게        (주)쇼박스    액션   2013-06-05   123       15세 관람가      장철수   
3     나는 공무원이다        (주)NEW   코미디   2012-07-12   101        전체 관람가      구자홍   
4         불량남녀  쇼박스(주)미디어플렉스   코미디   2010-11-04   108       15세 관람가      신근호   
..         ...           ...   ...          ...   ...           ...      ...   
595         해무        (주)NEW   드라마   2014-08-13   111      청소년 관람불가      심성보   
596       파파로티        (주)쇼박스   드라마   2013-03-14   127       15세 관람가      윤종찬   
597      살인의 강     (주)마운틴픽쳐스    공포   2010-09-30    99      청소년 관람불가      김대현   
598     악의 연대기     CJ 엔터테인먼트   느와르   2015-05-14   102       15세 관람가      백운학   
599        베를린     CJ 엔터테인먼트    액션   2013-01-30   120       15세 관람가      류승완   

     dir_prev_bfnum  dir_prev_num  num_staff  num_actor  box_off_num  
0              0.00             0         91          2        23398  
1        1161602.50             2        387          3      7072501  
2         220775.25             4        343          4      6959083  
3          23894.00             2         20          6       217866  
4              1.00             1        251          2       483387  
..              ...           ...        ...        ...          ...  
595         3833.00             1        510          7      1475091  
596       496061.00             1        286          6      1716438  
597            0.00             0        123          4         2475  
598            0.00             0        431          4      2192525  
599            0.00             0        363          5      7166532  

[600 rows x 12 columns]

In [ ]:
#checkcode
ensure_vals(globals(),'train')
@check_safety
def check(user_answer=_, user_train=train):

    c_point0 = (user_answer.shape == (600,12))

    if c_point0 :
        return True
    else:
        return False
check()

### Inst.

`fillna()`는 pandas 데이터프레임(DataFrame)에서 **결측치(missing value)를 특정 값으로** 채우는 함수입니다.  
해당 코드는 train 데이터프레임에서 결측치가 있는 셀을 **0으로 채운** 새로운 데이터프레임 train_fill을 반환합니다.  

<img width=585 src="https://drive.google.com/uc?id=1mUybEzOx6CgES6zpRYf6_AwcgZRQ8wsE">

이를 통해 결측치 처리 과정 중 하나인 **결측치 대체** 를 수행할 수 있습니다.  
결측치를 평균, 중앙값, 최빈값 등의 대표값으로 대체하거나, 특정한 값으로 채우는 방법을 선택할 수 있습니다.  

하지만, 모든 결측치를 0으로 대체하는 것이 항상 좋은 방법은 아닙니다.  
결측치 대체 방법은 데이터의 특성과 분석 목적에 따라 적절한 방법을 선택해야 합니다.  

### hint.
empty

### solution.
```python
train_fill = train.fillna(0)
train_fill
```

# 8. train_fill 데이터셋의 피처(feature)별 결측치 개수 확인하기 (3)

[문제 8]    

결측치가 제대로 제거되었는지 확인해 봐야겠죠?

피처(feature)별 결측치 개수의 합을 구해 보세요.


train_fill 데이터셋의 피처(feature)별 결측치 개수의 합을 구해 보세요.  
아래 빈칸을 채워보세요.   

In [ ]:
train_fill.___().___()

title             0
distributor       0
genre             0
release_time      0
time              0
screening_rat     0
director          0
dir_prev_bfnum    0
dir_prev_num      0
num_staff         0
num_actor         0
box_off_num       0
dtype: int64

In [ ]:
#checkcode
ensure_vals(globals(), 'train_dropna')
@check_safety
def check(check_train=train_dropna, user_answer=_):
    c_point0 = (user_answer == check_train.isnull().sum()).all()
    if c_point0:
        return True
    else:
        return False
check()

True

### Inst.



### Hint.
`isnull().sum()` 을 이용하면 train 데이터셋의 피처(feature)별 결측치의 수를 확인할 수 있습니다.

### Solution.
```python
train_fill.isnull().sum()
```

# 9. 결측치가 제거된 train_fill 데이터셋의 크기 확인하기

[문제 9]  
결측치들을 제거한 후의 행(row)과 열(column)의 크기를 확인해 봅시다.  
아래 빈칸을 채워주세요.


In [ ]:
train_fill.___

(600, 12)

In [ ]:
#checkcode
ensure_vals(globals(), 'train_fill')
@check_safety
def check(user_answer=_,user_answer_train=train):
    c_point0 = user_answer == user_answer_train.shape
    if c_point0:
        return True
    else:
        return False
check()

False

### Inst.

---

#### 결과 해석
위에서 출력된 값을 보면, 모든 피처(feature)의 결측치 개수가 **0으로 대체** 된 것을 알 수 있습니다.  
결측치를 0으로 대체한 train_fill 데이터프레임(DataFrame)은 **train 데이터프레임과 동일한 크기** 를 가지며,  
결측치를 0으로 대체한 부분만 다르게 처리된 것입니다.  

train 데이터셋의 행(row), 열(column)은 (600,12) 였으며, 결측치가 존재하는 330개의 행을 0으로 대체한 결과,  
train_fill 데이터프레임의 **행, 열은 여전히 (600,12)** 입니다.  

위 코드는 결측치를 처리하는 또 다른 방법으로, **결측치를 0으로 대체하여 데이터를 보존** 하는 것입니다.  
`fillna()` 함수를 사용하여 이를 구현하였습니다.  

결측치가 많은 경우에는 다른 값으로 대체하는 것이 더 적절할 수 있습니다.  
하지만 결측치가 일부 존재하거나 결측치의 비율이 적은 경우에는 0으로 대체하는 것도 하나의 대안일 수 있습니다.  

### Hint.
`shape` 함수를 사용해서 데이터프레임(dataframe)의 행(row)과 열(column)의 크기를 확인해 보세요.

### Solution.
```python
train_fill.shape
```